### ISSUE TITLE
Build once, but package N with different names 

### ISSUE THREAD

**Issue by lasote** [conan-io/conan#1798]
- Different conan packages from 1 build

**Comment by Aalmann** [comment]
Would be a very nice feature. I have a good use case: 

The underlying build-system (let's call it maven-tycho ;-) ) produces binaries (let's call it Eclipse-RCP-Application) for different platforms (Windows, Linux, Mac). Now I have 3 ZIP files created by Maven-tycho for each platform. But I'm unable to create 3 different packages, so that the consumer only retrieves the required package (for Windows, Linux, Mac).
Yes sounds weird to use Conan for building Eclipse-RCP-Apps, but so we have a complete universal build environment and I don't have to care about CI configuration.
Current workaround: Build the packages on different machines.

**Comment by lasote** [comment]
But if the app is the same but you have binaries for 3 different configurations is totally ok. To have 1 recipe reference and N binary packages (1 for each different config). It is the regular Conan operation mode. Am I missing something?

**Comment by lasote** [comment]
By the way, this "build once" and package many with different names can be done creating a package for the whole configurations and using it as a build require of N different packages with different names so I think it could be closed now.

**Comment by Aalmann** [comment]
No. The difference, especially for my use case, is the count `conan create` call itself. 
Normally we are using Jenkins with Matrix Jobs to build for different platforms with the same job but different settings for each platform. This ends in "platform count" calls of conan create for the same recipe with the same name but different settings (os, arch, etc.). Thats exactly what you describe,

But when I build an Eclipse-RCP app with maven-tycho, maven itself creates zip files for different platforms. That means, one Jenkins job on one platform with one `conan create` call produces more than one different packages for each (maven) platform. So in my case I have 3 zip files and I want to be able to create with this one call 3 different packages from one build directory.

Another usecase could be to create artefacts for multiple archs (x86, x86_64 and ARM) with one build or build call within the build() method, assuming the used build system is able to do so. Within Visual Studio I'm able to switch between different archs. And it is also possible to do so at command line.

**Comment by memsharded** [comment]
@Aalmann I think the above suggested approach can work:

- Create a package lets say ``PkgAll/1.0@...`` that contains the call to Eclipse-RCP and build and packages the 3 zip files. 
- You can manage to do it multi-plaform: just remove the settings, like ``"os"`` that do not affect. You can leave, ``"arch"`` if it does
- Create a package named ``Pkg/0.1@...`` that "build-requires" ``Pkg/0.1@...``
- Don't remove the settings in this package.
- You can use the ``imports()`` feature to make a copy of the zip you want. Make sure to add the ``keep_imports=True``.
- Re-package() just the binaries for that specific system.

Maybe this section could help: http://docs.conan.io/en/latest/devtools/running_packages.html?highlight=re-packaging#runtime-packages-and-re-packaging
Please tell me if this helps. Thanks!

**Comment by Aalmann** [comment]
@memsharded 
I thought about that approach, but it ends with at least 4 `conan create` calls (over all ci-jobs) for 3 different platforms. Number 1 ist for platform independent PkgAll. After that I have to upload the package to be able to get the zip files on the other platforms (downstream matrix job) or I could leave everything local and call `conan create` directly after PkgAll creation for the 3 different platforms. 
The result: 1 simple recipe and 1 additional recipe (to maintain) and 4 `conan create` calls.

On the other hand I could also use one recipe in a matrix job on 3 different platforms to call `conan create`. This will end up with 3 maven calls but also 3 suitable packages for each platform. The only thing is that just one platform has to be able to deploy the results to maven repo.
The result: 1 more complex recipe and 3 `conan create` calls.

With the build once but package many approach there would be 1 (probably a little bit complex) recipe and 1 `conan create` call.


Btw.: For those who ask them-self why we are using conan to call a maven build. Conan is used for all our other packages (C/C++), so it is installed/present at our build slaves which are not connected to the internet but to an artifact repo. And conan can be used to setup the build toolchain from scratch (in this case maven and jdk).

**Comment by memsharded** [comment]
Thanks @Aalmann for the feedback.
I still can't figure out how to create 3 different packages with a single ``conan create`` call. What would be the user interface to specify that? 

Maybe in terms of recipe simplicity, the approach could be to have just 1 recipe, use the local flow to ``install + build``, then call 3 times the ``export-pkg`` to create the correct final package. Have you considered this possibility?

**Comment by Aalmann** [comment]
@memsharded 
counter question: What' your picture of Build once package many?

My idea/wish of a single `conan create`:
```
> conan create . user/channel -s os=ANY -s compiler=ANY -s build_type=ANY...
or
> conan create . user/channel -s os=None -s compiler=None -s build_type=None ...
or
> conan create . user/channel -s os="Windows","Linux"-s compiler="ANY" -s build_type="ANY" -s arch="x86","x86_64" ...
``` 
can be used within the recipe like this:
```
...
settings = { "os": ["Windows", "Linux"], "compiler": "ANY", "arch": ["x86", "x86_64"], "build_type": "ANY"}
...
def build(self):
    # only required if the build system itself is unable to directly build for multiple combinations (like multiple archs, etc.)
    for set_opt_combination in tools.get_combinations(self):
        if set_opt_combination.get("os") == "Windows":
             # build something for Windows
        elif set_opt_combination.get("os") == "Linux":
            # build something for Linux
        else:
           # else
    # or just something like, which will create multiple packages with the maven layer
    self.run("mvn ...")

def package(self):
    for set_opt_combination in tools.get_combinations(self):
        package_folder = tools.get_package_folder(set_opt_combination)
        if set_opt_combination.get("os") == "Windows" and set_opt_combination.get("arch") == "x86":
             # package for Windows x86
            self.copy(... package_folder=package_folder)
        elif set_opt_combination.get("os") == "Linux" and set_opt_combination.get("arch") == "x86_64":
            # package for Linux x86_64
            self.copy(... package_folder=package_folder)
        else:
           # else
```

**Comment by memsharded** [comment]
> > conan create . user/channel -s os=ANY -s compiler=ANY -s build_type=ANY

But that doesn't create packages with different names. In any case, that will create different binaries for the same package, for different configurations. I am not sure this is the scope of this issue.

Regarding your syntax, this will create an explosion of combinations over all the possibilities of settings.yml. The settings declared in recipes are just a "constraint" over the input, but the command input operates over the overall settings definition.

My view is that the relation 1 create => 1 package works very well, it is clear what it does and is understandable, easier to debug, etc. Also, using settings is not recommended for production, the way to go is using profiles. So I think the correct approach would be:

```bash
$ conan create . user/testing -pr=myprofile_win
$ conan create . user/testing -pr=myprofile_...
$ conan create . user/testing -pr=myprofile_...
```

And what you want is that such process is efficient, that it doesn't build three times the same thing? Then you definitely want the ``build_id()`` functionality, which does exactly that. Please have a look is it what you are looking for: http://docs.conan.io/en/latest/creating_packages/package_approaches.html#n-configs-1-build-n-packages ?

If the above 3 creates are bothering you, why not just put them in a .bat, .sh or .py script in the root of the package? Thanks for your feedback.

**Comment by Aalmann** [comment]
I don't want to create packages with different names. I want to create N packages after 1 build using the same recipe.

Yes, "ANY" could end in an explosion. But with `-s os="Windows","Linux"` ist could work.

IMHO using settings instead of profiles is the more obvious way. First, the profiles have to be shared and there is currently no concept for that. Second, using the command line is more clear (except it is more to write). The profile name itself does not clarify whats in there.

The 3 calls don't bother me. I'm using it that way right now. But it would be nice to get the opportunity to split the build result into several packages. This could be useful for maven builds (like in my case) or to extract and separate an external ZIP/tgz/installer/whatever file which contains debug and release artifacts into different packages without downloading and extracting the external package twice.

**Comment by Aalmann** [comment]
Thanks for the hint with `conan config install`. Didn't recognize this feature. I will check it.

I know what `build_id()` does. And that is not the point.
Btw. I have scrolled up this issue and found that you are talking about `Build once, but package N with different names` (which is the title) and I'm just talking about `Build once, but package N with the same name` (which is not the title). Sorry about that and the confusion, I have to read carefully next time.

However: What I mean or try to describe is, that I only want to run `source()` and `build()` once and `package()` N-times. That's why I mentioned a single `conan create` call.
UseCase1: Build an Eclipse-RCP with Maven-Tycho. I want to checkout once, I want to build once and I want to package N-times. Result: N packages for N os, arch, etc. combination.
Usecase2: Extracting a "non-conan" Installer/Zip/Tgz/Whatever file which contains both build_types (Release, Debug). I want to download the file once, I want to extract the file once, but I want to package it twice. Result: 2 packages for each build_type.

I hope you got an idea of what I want to achieve, even if it's not the right issue for that use case.

**Comment by memsharded** [comment]
This is a stale issue.

I still think that the ``build_id()`` does exactly that. When a recipe defines a ``build_id()`` method:

- Even if you run multiple ``conan create``, the ``source()`` method will be called only once
- Even if you run multiple ``conan create``, the ``build()`` method will be called only once, only the first ``conan create``, the following ``conan create`` will not call ``build()`` method at all.
- For the multiple ``conan create``, the ``package()`` method will be called once per configuration, reused the same "build" folder that was built exactly once.

So I am closing this issue as solved, but please let me know if there are further questions. Best would be to create a new ticket.

Regarding the original OP, building once, but package with **different** package names, this can be done with the ``conan export-pkg --name=xxxx`` if you build locally, but in the cache it is not a planned feature, so this ticket can be closed too.
